File: modules/jobs.py
"""Durable, DB-backed job queue.

Replaces the daemon-thread spawns scattered through app.py and pipeline.py
so pipeline runs, server-create flows, and full-delete teardowns survive a
Flask restart instead of dying mid-step.

DESIGN
  - Single in-process worker thread polls the `jobs` table every POLL_INTERVAL,
    claims one queued job at a time, runs the registered handler, updates
    status. The QUEUE is durable; the worker is still in-process. Multi-process
    workers can come in v2.
  - Handlers are registered by string kind ("pipeline.full", etc.) at module
    load time in app.py and dispatched by the worker.
  - Boot recovery: any row stuck in 'running' from a prior process gets reset
    to 'queued' (if attempts left) or 'failed' (if exhausted).
  - Cancellation: the existing per-domain cancel_requested flag still handles
    in-flight pipeline cancels via _check_cancel() at step boundaries.
    cancel_queued_job() flips a queued-but-not-yet-running job to 'canceled'.

LIMITATIONS (acknowledged)
  - Single worker → jobs serialize. Today's daemon threads ran in parallel.
    Tradeoff for durability + simplicity. Worker pool is a v2 concern.
  - No retry backoff. attempt_count gates against runaway retries via
    max_attempts (default 1).
  - FIFO by id; no priorities.
"""

from __future__ import annotations

import json
import os
import threading
import time
import traceback
from typing import Callable

from database import get_db


# Registered handlers — populated by register_handler() at import time in app.py.
_handlers: dict[str, Callable[[dict], None]] = {}

# Worker pool. Each thread runs the same _worker_loop, claiming one job at a
# time. Pool size is configurable via start_worker(num_workers=N) — default 4.
# Different jobs (different domains) run in parallel; the per-domain
# _inflight_domains set in pipeline.py still serializes same-domain runs.
#
# Why threads (not multiprocess, not asyncio): every long-running operation in
# this codebase is I/O-bound (HTTP, SSH, sqlite, LLM). The GIL releases on I/O,
# so threads achieve real parallelism for our workload. Multiprocess would
# require picklable handlers + cross-process domain locking; asyncio would
# require rewriting every requests/paramiko/sqlite3 call. Threads are the
# right ergonomic fit.
_worker_threads: list[threading.Thread] = []
_worker_stop = threading.Event()
_worker_lock = threading.Lock()
_worker_id_base = f"pid-{os.getpid()}"

POLL_INTERVAL = 2.0
# Configurable via SSR_JOB_WORKERS env var (set in /etc/ssr.env on prod).
# Bumping this raises pipeline parallelism but also raises the rate at which
# this process hits the CF / DO / SA APIs. Tune to your account quotas.
try:
    DEFAULT_WORKERS = max(1, int(os.environ.get("SSR_JOB_WORKERS", "4")))
except ValueError:
    DEFAULT_WORKERS = 4


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def register_handler(kind: str, fn: Callable[[dict], None]) -> None:
    """Register a handler for a job kind. 1:1 with kinds — re-registering raises."""
    if kind in _handlers:
        raise ValueError(f"handler for {kind!r} already registered")
    _handlers[kind] = fn


def enqueue_job(kind: str, payload: dict, max_attempts: int = 1) -> int:
    """Insert a queued job. Returns the new row id.

    Does NOT validate that `kind` has a handler — handlers can register after
    enqueue (boot sequencing). Dispatch-time errors mark the job failed.
    """
    now = time.time()
    conn = get_db()
    try:
        cur = conn.execute(
            """INSERT INTO jobs
                 (kind, payload_json, status, attempt_count, max_attempts,
                  created_at, updated_at)
               VALUES (?, ?, 'queued', 0, ?, ?, ?)""",
            (kind, json.dumps(payload), max_attempts, now, now),
        )
        conn.commit()
        return cur.lastrowid
    finally:
        conn.close()


def get_job(job_id: int) -> dict | None:
    conn = get_db()
    try:
        row = conn.execute("SELECT * FROM jobs WHERE id = ?", (job_id,)).fetchone()
        return dict(row) if row else None
    finally:
        conn.close()


def list_jobs(status: str | None = None, kind: str | None = None,
              limit: int = 50) -> list[dict]:
    where = []
    args: list = []
    if status:
        where.append("status = ?"); args.append(status)
    if kind:
        where.append("kind = ?"); args.append(kind)
    sql = "SELECT * FROM jobs"
    if where:
        sql += " WHERE " + " AND ".join(where)
    sql += " ORDER BY id DESC LIMIT ?"
    args.append(limit)
    conn = get_db()
    try:
        return [dict(r) for r in conn.execute(sql, args).fetchall()]
    finally:
        conn.close()


def cancel_queued_job(job_id: int) -> bool:
    """Mark a queued job as canceled. Returns True if it transitioned, False
    if it wasn't queued (already running/done/canceled/failed)."""
    now = time.time()
    conn = get_db()
    try:
        cur = conn.execute(
            """UPDATE jobs
                  SET status = 'canceled', updated_at = ?
                WHERE id = ? AND status = 'queued'""",
            (now, job_id),
        )
        conn.commit()
        return cur.rowcount > 0
    finally:
        conn.close()


def recover_orphans() -> int:
    """Boot-time recovery. Any row left in 'running' from a prior process is
    orphaned (its worker is gone). Reset to 'queued' if attempts remain, else
    mark 'failed'. Returns total rows touched.
    """
    now = time.time()
    conn = get_db()
    try:
        # Mark exhausted-attempt orphans failed first (so the requeue UPDATE
        # below doesn't pick them up).
        cur1 = conn.execute(
            """UPDATE jobs
                  SET status = 'failed',
                      last_error = COALESCE(last_error, '')
                                || ' | orphaned: process restarted mid-run',
                      locked_by = NULL,
                      locked_at = NULL,
                      updated_at = ?
                WHERE status = 'running'
                  AND attempt_count >= max_attempts""",
            (now,),
        )
        cur2 = conn.execute(
            """UPDATE jobs
                  SET status = 'queued',
                      locked_by = NULL,
                      locked_at = NULL,
                      updated_at = ?
                WHERE status = 'running'""",
            (now,),
        )
        conn.commit()
        return cur1.rowcount + cur2.rowcount
    finally:
        conn.close()


# ---------------------------------------------------------------------------
# Worker internals
# ---------------------------------------------------------------------------

def _claim_one(worker_id: str = None) -> dict | None:
    """Grab the oldest queued job. Returns the post-claim row dict (with
    bumped attempt_count + 'running' status) or None if queue is empty.

    With multiple workers, the `AND status = 'queued'` guard on the UPDATE
    + cur.rowcount check ensures only one worker can claim a given row.
    SELECT-then-UPDATE is non-atomic but the guard makes the loser's UPDATE
    a no-op — they just retry next poll.
    """
    if worker_id is None:
        worker_id = _worker_id_base
    now = time.time()
    conn = get_db()
    try:
        row = conn.execute(
            """SELECT id FROM jobs
                WHERE status = 'queued'
                ORDER BY id
                LIMIT 1"""
        ).fetchone()
        if not row:
            return None
        cur = conn.execute(
            """UPDATE jobs
                  SET status = 'running',
                      attempt_count = attempt_count + 1,
                      locked_by = ?,
                      locked_at = ?,
                      updated_at = ?
                WHERE id = ? AND status = 'queued'""",
            (worker_id, now, now, row["id"]),
        )
        conn.commit()
        if cur.rowcount == 0:
            # Lost the race to another worker — try again next tick.
            return None
        claimed = conn.execute(
            "SELECT * FROM jobs WHERE id = ?", (row["id"],)
        ).fetchone()
        return dict(claimed) if claimed else None
    finally:
        conn.close()


def _finish(job_id: int, status: str, error: str | None = None) -> None:
    now = time.time()
    conn = get_db()
    try:
        conn.execute(
            """UPDATE jobs
                  SET status = ?,
                      last_error = ?,
                      locked_by = NULL,
                      locked_at = NULL,
                      updated_at = ?
                WHERE id = ?""",
            (status, error, now, job_id),
        )
        conn.commit()
    finally:
        conn.close()


def _requeue_for_retry(job_id: int, error: str) -> None:
    now = time.time()
    conn = get_db()
    try:
        conn.execute(
            """UPDATE jobs
                  SET status = 'queued',
                      last_error = ?,
                      locked_by = NULL,
                      locked_at = NULL,
                      updated_at = ?
                WHERE id = ?""",
            (error, now, job_id),
        )
        conn.commit()
    finally:
        conn.close()


def _run_one(job: dict) -> None:
    handler = _handlers.get(job["kind"])
    if handler is None:
        _finish(job["id"], "failed",
                f"No handler registered for kind={job['kind']!r}")
        return
    try:
        payload = json.loads(job["payload_json"])
    except json.JSONDecodeError as e:
        _finish(job["id"], "failed", f"Bad payload JSON: {e}")
        return
    try:
        handler(payload)
        _finish(job["id"], "done")
    except Exception as e:
        tb = traceback.format_exc()[:4000]
        err = f"{type(e).__name__}: {e}\n{tb}"
        # attempt_count was bumped at claim time, so it now reflects this attempt.
        if job["attempt_count"] < job["max_attempts"]:
            _requeue_for_retry(job["id"], err)
        else:
            _finish(job["id"], "failed", err)


def _worker_loop(worker_id: str) -> None:
    while not _worker_stop.is_set():
        try:
            job = _claim_one(worker_id=worker_id)
        except Exception:
            # If we can't even claim (DB lock contention, schema mismatch), back
            # off and retry rather than spin a tight failure loop.
            _worker_stop.wait(POLL_INTERVAL * 2)
            continue
        if job is None:
            _worker_stop.wait(POLL_INTERVAL)
            continue
        _run_one(job)


def start_worker(num_workers: int = DEFAULT_WORKERS) -> None:
    """Idempotently start a pool of N worker threads. Safe to call multiple
    times — extra calls are no-ops while the existing pool is alive.

    Each worker has a distinct worker_id ('pid-1234-w0', 'pid-1234-w1', ...)
    written to jobs.locked_by so we can tell which worker is holding which
    row in the (rare) case of contention or stuck jobs.
    """
    global _worker_threads
    with _worker_lock:
        alive = [t for t in _worker_threads if t.is_alive()]
        if alive:
            return
        _worker_stop.clear()
        _worker_threads = []
        for i in range(num_workers):
            wid = f"{_worker_id_base}-w{i}"
            t = threading.Thread(
                target=_worker_loop,
                args=(wid,),
                daemon=True,
                name=f"ssr-job-worker-{i}",
            )
            t.start()
            _worker_threads.append(t)


def stop_worker(timeout: float = 5.0) -> None:
    """Stop all workers. Used in tests. Currently-running handlers, if any,
    finish — they aren't interrupted."""
    _worker_stop.set()
    for t in _worker_threads:
        t.join(timeout=timeout)

File: template_v0/lib/jobs.ts
/**
 * Durable, DB-backed job queue — Node port of modules/jobs.py.
 *
 * Same `jobs` table the Flask side uses. Workers poll every 2s, claim
 * one row at a time, dispatch to a registered handler, write status +
 * last_error back. Boot recovery resets stuck 'running' rows.
 *
 * Single in-process pool of N async workers (default 4). Tunable via
 * SSR_JOB_WORKERS env var. Async/await for the loop, but each handler
 * runs to completion before the worker claims the next row — same
 * single-job-per-worker semantics as the Python version.
 *
 * The Flask job worker pool ALSO polls this same table. If both are
 * running side-by-side, whichever pool's worker wins the
 * `AND status='queued'` UPDATE race claims the row. For now we
 * recommend running ONE worker pool at a time to avoid double-pickup
 * confusion (set SSR_JOB_WORKERS=0 on whichever side you want quiet).
 */
import { all, getDb, one, run } from "./db"

export type JobHandler = (payload: Record<string, unknown>) => void | Promise<void>

export interface JobRow {
  id: number
  kind: string
  payload_json: string
  status: string
  attempt_count: number
  max_attempts: number
  locked_by: string | null
  locked_at: number | null
  last_error: string | null
  created_at: number
  updated_at: number
}

const POLL_INTERVAL_MS = 2000

// HMR-safe state. Next dev / Turbopack re-evaluates this module on edits;
// keeping these on globalThis means each re-eval reuses the existing pool
// instead of spawning a fresh set of workers while the old loops keep
// polling. Fixes the multi-pool leak that compounded the CPU spike.
declare global {
  // eslint-disable-next-line no-var
  var __ssrJobPool: { started: boolean; stop: boolean; threads: Promise<void>[] } | undefined
  // eslint-disable-next-line no-var
  var __ssrJobHandlers: Map<string, JobHandler> | undefined
}
const pool = (globalThis.__ssrJobPool ??= { started: false, stop: false, threads: [] })
const handlers = (globalThis.__ssrJobHandlers ??= new Map<string, JobHandler>())

const WORKER_ID_BASE = `node-pid-${process.pid}`

function defaultWorkers(): number {
  const raw = process.env.SSR_JOB_WORKERS
  if (!raw) return 4
  const n = Number.parseInt(raw, 10)
  if (!Number.isFinite(n) || n < 0) return 4
  return Math.max(1, n)
}

export function registerHandler(kind: string, fn: JobHandler): void {
  // Idempotent re-register: HMR re-runs instrumentation, which re-imports
  // pipeline/cf-bulk/etc. and calls registerHandler again. Throwing here
  // would spam errors on every dev edit; replacing the entry is fine since
  // the new function is the latest source.
  handlers.set(kind, fn)
}

export function enqueueJob(kind: string, payload: unknown, maxAttempts = 1): number {
  const now = Date.now() / 1000
  const result = run(
    `INSERT INTO jobs (kind, payload_json, status, attempt_count, max_attempts, created_at, updated_at)
       VALUES (?, ?, 'queued', 0, ?, ?, ?)`,
    kind,
    JSON.stringify(payload),
    maxAttempts,
    now,
    now,
  )
  return Number(result.lastInsertRowid)
}

export function getJob(id: number): JobRow | undefined {
  return one<JobRow>("SELECT * FROM jobs WHERE id = ?", id)
}

export function listJobs(opts: { status?: string; kind?: string; limit?: number } = {}): JobRow[] {
  const where: string[] = []
  const args: (string | number)[] = []
  if (opts.status) { where.push("status = ?"); args.push(opts.status) }
  if (opts.kind)   { where.push("kind = ?");   args.push(opts.kind) }
  let sql = "SELECT * FROM jobs"
  if (where.length) sql += " WHERE " + where.join(" AND ")
  sql += " ORDER BY id DESC LIMIT ?"
  args.push(opts.limit ?? 50)
  return all<JobRow>(sql, ...args)
}

export function recoverOrphans(): number {
  const now = Date.now() / 1000
  // attempt_count is incremented on FAILURE (in runOne), not on CLAIM,
  // so a crashed worker leaves the row at its pre-claim count. The check
  // below correctly fails only jobs that have actually exhausted their
  // recorded attempts; everything else gets requeued for another try.
  const failed = run(
    `UPDATE jobs
        SET status = 'failed',
            last_error = COALESCE(last_error, '') || ' | orphaned: process restarted mid-run',
            locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE status = 'running' AND attempt_count >= max_attempts`,
    now,
  )
  // Mark pipeline.full orphans whose domain is already at a success status
  // as 'done' instead of requeueing — otherwise a worker killed AFTER its
  // pipeline finished but BEFORE finishJob() runs would re-fire the whole
  // pipeline (step 4 hits Spaceship, step 5 polls CF for 2 min) on a
  // domain that's already serving. Inline the success-status list rather
  // than importing from status-taxonomy to avoid a cycle.
  const skippedDone = run(
    `UPDATE jobs
        SET status = 'done',
            last_error = COALESCE(last_error, '') || ' | orphan recovered: domain already at success',
            locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE status = 'running'
        AND kind = 'pipeline.full'
        AND EXISTS (
          SELECT 1 FROM domains d
          WHERE d.domain = json_extract(jobs.payload_json, '$.domain')
            AND d.status IN ('hosted', 'live', 'completed')
        )`,
    now,
  )
  // Requeue the rest.
  const requeued = run(
    `UPDATE jobs
        SET status = 'queued', locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE status = 'running'`,
    now,
  )
  return Number(failed.changes) + Number(skippedDone.changes) + Number(requeued.changes)
}

/**
 * Soft variant of `recoverOrphans` for the in-process stuck-job sweeper.
 * Only requeues rows that have been `running` long enough to be considered
 * stuck (default 30 min). Used by the scheduled sweeper so a worker that
 * silently dies — e.g. an upstream `await` that never resolves and isn't
 * caught by Node's process-level handler — gets its job back to the queue
 * within bounded time instead of waiting for the next process restart.
 */
export function recoverStuckJobs(stuckAfterSeconds = 1800): number {
  const now = Date.now() / 1000
  const cutoff = now - stuckAfterSeconds
  const failed = run(
    `UPDATE jobs
        SET status = 'failed',
            last_error = COALESCE(last_error, '') || ' | stuck: locked > ' || ? || 's, attempts exhausted',
            locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE status = 'running'
        AND locked_at IS NOT NULL AND locked_at < ?
        AND attempt_count >= max_attempts`,
    stuckAfterSeconds,
    now,
    cutoff,
  )
  const requeued = run(
    `UPDATE jobs
        SET status = 'queued',
            last_error = COALESCE(last_error, '') || ' | stuck: locked > ' || ? || 's, requeued',
            locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE status = 'running'
        AND locked_at IS NOT NULL AND locked_at < ?`,
    stuckAfterSeconds,
    now,
    cutoff,
  )
  return Number(failed.changes) + Number(requeued.changes)
}

function claimOne(workerId: string): JobRow | null {
  const now = Date.now() / 1000
  const db = getDb()
  // SELECT-then-UPDATE with status guard. SQLite serializes writes; the
  // guard makes the loser's UPDATE a no-op (rowcount=0).
  // attempt_count is intentionally NOT incremented here — it ticks up only
  // on actual failure, so a worker that crashes mid-handler doesn't burn
  // an attempt. Recovery is then accurate (see recoverOrphans).
  const row = db.prepare(
    `SELECT id FROM jobs WHERE status = 'queued' ORDER BY id LIMIT 1`,
  ).get() as { id: number } | undefined
  if (!row) return null
  const updated = db.prepare(
    `UPDATE jobs
        SET status = 'running',
            locked_by = ?, locked_at = ?, updated_at = ?
      WHERE id = ? AND status = 'queued'`,
  ).run(workerId, now, now, row.id)
  if (Number(updated.changes) === 0) return null
  return one<JobRow>("SELECT * FROM jobs WHERE id = ?", row.id) ?? null
}

function finishJob(id: number, status: string, error: string | null = null): void {
  const now = Date.now() / 1000
  run(
    `UPDATE jobs
        SET status = ?, last_error = ?, locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE id = ?`,
    status,
    error,
    now,
    id,
  )
}

function failJobIncrement(id: number, error: string, newCount: number): void {
  const now = Date.now() / 1000
  run(
    `UPDATE jobs
        SET status = 'failed', attempt_count = ?, last_error = ?,
            locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE id = ?`,
    newCount,
    error,
    now,
    id,
  )
}

function requeueForRetry(id: number, error: string, newCount: number): void {
  const now = Date.now() / 1000
  run(
    `UPDATE jobs
        SET status = 'queued', attempt_count = ?, last_error = ?,
            locked_by = NULL, locked_at = NULL, updated_at = ?
      WHERE id = ?`,
    newCount,
    error,
    now,
    id,
  )
}

async function runOne(job: JobRow): Promise<void> {
  const fn = handlers.get(job.kind)
  if (!fn) {
    finishJob(job.id, "failed", `No handler registered for kind=${JSON.stringify(job.kind)}`)
    return
  }
  let payload: unknown
  try {
    payload = JSON.parse(job.payload_json)
  } catch (e) {
    finishJob(job.id, "failed", `Bad payload JSON: ${(e as Error).message}`)
    return
  }
  try {
    await fn(payload as Record<string, unknown>)
    finishJob(job.id, "done")
  } catch (e) {
    const err = `${(e as Error).name}: ${(e as Error).message}`
    // Increment attempt_count on actual failure (not on claim) so a crashed
    // worker doesn't burn an attempt. After this increment, if the row has
    // recorded N >= max_attempts failures, it's terminal; otherwise requeue.
    const newCount = job.attempt_count + 1
    if (newCount < job.max_attempts) {
      requeueForRetry(job.id, err, newCount)
    } else {
      failJobIncrement(job.id, err, newCount)
    }
  }
}

async function workerLoop(workerId: string): Promise<void> {
  while (!pool.stop) {
    let job: JobRow | null = null
    try {
      job = claimOne(workerId)
    } catch (e) {
      // Don't tight-loop on DB errors.
      await sleep(POLL_INTERVAL_MS * 2)
      continue
    }
    if (!job) {
      await sleep(POLL_INTERVAL_MS)
      continue
    }
    await runOne(job)
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms))
}

export function startPool(numWorkers = defaultWorkers()): void {
  if (pool.started) return
  pool.started = true
  pool.stop = false
  recoverOrphans()
  for (let i = 0; i < numWorkers; i++) {
    const id = `${WORKER_ID_BASE}-w${i}`
    pool.threads.push(workerLoop(id))
  }
  startStuckJobSweeper()
}

export async function stopPool(): Promise<void> {
  pool.stop = true
  await Promise.all(pool.threads)
  pool.threads.length = 0
  pool.started = false
  if (globalThis.__ssrStuckJobTimer) {
    clearInterval(globalThis.__ssrStuckJobTimer)
    globalThis.__ssrStuckJobTimer = undefined
  }
}

declare global {
  // eslint-disable-next-line no-var
  var __ssrStuckJobTimer: ReturnType<typeof setInterval> | undefined
}

const STUCK_SWEEP_INTERVAL_MS = 5 * 60 * 1000

/**
 * In-process sweeper: every 5 min, walks the jobs table for `running` rows
 * locked > 30 min and either requeues them (if attempts remain) or marks
 * them failed. Closes the gap where a worker silently wedges (await that
 * never resolves, sync infinite loop, GC pause) and the row sits running
 * forever — recoverOrphans only runs at boot.
 */
function startStuckJobSweeper(): void {
  if (globalThis.__ssrStuckJobTimer) return
  if (process.env.NODE_ENV === "test") return
  if (process.env.SSR_JOB_SWEEPER === "0") return
  const stuckAfter = parseStuckAfterSeconds()
  globalThis.__ssrStuckJobTimer = setInterval(() => {
    try {
      const n = recoverStuckJobs(stuckAfter)
      if (n > 0) {
        // eslint-disable-next-line no-console
        console.log(`[jobs] stuck-job sweeper recovered ${n} row(s)`)
      }
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error("[jobs] stuck-job sweeper failed:", e)
    }
  }, STUCK_SWEEP_INTERVAL_MS)
  globalThis.__ssrStuckJobTimer.unref?.()
}

function parseStuckAfterSeconds(): number {
  const n = parseInt(process.env.SSR_JOB_STUCK_AFTER_S || "1800", 10)
  return Number.isFinite(n) && n >= 60 ? n : 1800
}
